physical/raft/raft_autopilot.go

```diff
@@ -263,30 +268,37 @@ type Delegate struct {
 	*RaftBackend
 
 	// dl is a lock dedicated for guarding delegate's fields
-	dl               sync.RWMutex
-	inflightRemovals map[raft.ServerID]bool
-	emptyVersionLogs map[raft.ServerID]struct{}
+	dl                 sync.RWMutex
+	inflightRemovals   map[raft.ServerID]bool
+	emptyVersionLogs   map[raft.ServerID]struct{}
+	permanentNonVoters map[raft.ServerID]bool
 }
 
 func NewDelegate(b *RaftBackend) *Delegate {
 	return &Delegate{
-		RaftBackend:      b,
-		inflightRemovals: make(map[raft.ServerID]bool),
-		emptyVersionLogs: make(map[raft.ServerID]struct{}),
+		RaftBackend:        b,
+		inflightRemovals:   make(map[raft.ServerID]bool),
+		emptyVersionLogs:   make(map[raft.ServerID]struct{}),
+		permanentNonVoters: make(map[raft.ServerID]bool),
 	}
 }
 
 // AutopilotConfig is called by the autopilot library to know the desired
 // autopilot configuration.
 func (d *Delegate) AutopilotConfig() *autopilot.Config {
+	// Always fetch non-voters before returning the config
+	err := d.FetchNonVoters()
+	if err != nil {
+		d.logger.Error("failed to fetch non-voters", "error", err)
+	}
 	d.l.RLock()
```

Comment:
What is the ordering/locking requirements w.r.t. `d.l` vs `d.dl`?

Reply:
Good question. `d.l` is the lock which the delegate inherits from it's `RaftBackend` instance, `d.dl` is the lock for it's own fields. Which means, I am missing a read lock for the non-voter map here :+1: